nbconvert/preprocessors/execute.py

```diff
@@ -295,40 +240,44 @@ def setup_preprocessor(self, nb, resources, km=None):
         resources : dictionary
             Additional resources used in the conversion process.
         """
+        # Nbconvert can be installed without these, so do the imports only
+        # when the preprocessor is being used.
+        from jupyter_kernel_mgmt.discovery import KernelFinder
+        from jupyter_kernel_mgmt.client import BlockingKernelClient
+
         path = resources.get('metadata', {}).get('path', '') or None
         self.nb = nb
         # clear display_id map
         self._display_id_map = {}
 
-        if km is None:
-            self.km, self.kc = self.start_new_kernel(cwd=path)
-            try:
-                # Yielding unbound args for more easier understanding and downstream consumption
-                yield nb, self.km, self.kc
-            finally:
-                self.kc.stop_channels()
-                self.km.shutdown_kernel(now=self.shutdown_kernel == 'immediate')
-
-                for attr in ['nb', 'km', 'kc']:
-                    delattr(self, attr)
+        kf = KernelFinder.from_entrypoints()
+        if self.kernel_name:
+            kernel_name = self.kernel_name
+            if '/' not in kernel_name:
+                kernel_name = 'spec/' + kernel_name
         else:
-            self.km = km
-            if not km.has_kernel:
-                km.start_kernel(extra_arguments=self.extra_arguments, **kwargs)
-            self.kc = km.client()
-
-            self.kc.start_channels()
-            try:
-                self.kc.wait_for_ready(timeout=self.startup_timeout)
-            except RuntimeError:
-                self.kc.stop_channels()
-                raise
-            self.kc.allow_stdin = False
             try:
-                yield nb, self.km, self.kc
-            finally:
-                for attr in ['nb', 'km', 'kc']:
-                    delattr(self, attr)
+                kernel_name = 'spec/' + nb.metadata.get('kernelspec', {})[
+                    'name']
+            except KeyError:
+                kernel_name = 'pyimport/kernel'
+
+        self.log.info("Launching kernel %s to execute notebook" % kernel_name)
+        conn_info, self.km = kf.launch(kernel_name, cwd=path)
```

Comment:
To address the async jupyter_kernel_mgmt changes, the only change required to get things working is replacing this line with the following (along with accompanying `import asyncio` statement)...
```python
conn_info, self.km = asyncio.get_event_loop().run_until_complete(kf.launch(kernel_name, cwd=path))
```

However, what is strange (and independent of this change) is that when I run nbconvert via a debugger (pycharm), I get the following exception: 
```
  File "/Users/kbates/repos/oss/gateway-experiments/nbconvert/nbconvert/exporters/exporter.py", line 315, in _preprocess
    nbc, resc = preprocessor(nbc, resc)
  File "/Users/kbates/repos/oss/gateway-experiments/nbconvert/nbconvert/preprocessors/base.py", line 47, in __call__
    return self.preprocess(nb, resources)
  File "/Users/kbates/repos/oss/gateway-experiments/nbconvert/nbconvert/preprocessors/execute.py", line 318, in preprocess
    nb.metadata['language_info'] = info_dict['language_info']
  File "/opt/anaconda3/envs/kernel-mgmt-dev/lib/python3.6/contextlib.py", line 92, in __exit__
    raise RuntimeError("generator didn't stop")
```
Since this occurs in either case, I figured its a side-effect of having a `yield` in a contextmanager - but I'm also a relative novice in python - so that may not be what's going on.  When run from the command line, also in both cases, everything works.

These changes also ignore the case where the kernel manager instance could be passed as an argument.  We should remove that parameter if that's no longer applicable.

Reply:
@takluyver -
> Since this occurs in either case, I figured its a side-effect of having a yield in a contextmanager - but I'm also a relative novice in python - so that may not be what's going on. When run from the command line, also in both cases, everything works.

Well, obviously a `yield` in a context manager is correct (yes, I've learned a little), but I suspect something is getting side-affected by the async changes in conjunction with the content manager.  If I break things down to not use a context manager, `nbconvert --execute` no longer encounters the "generator didn't stop" issue.

I've gone ahead and created a [PR against your branch](https://github.com/takluyver/nbconvert/pull/1) in case you want to take these changes.  If not, just close the PR.  Thanks